File: lekcje/01_algorytmy_i_zlozonosc/przyklady/sortowanie.cs
using System;
using System.Collections.Generic;

namespace SortingApplication
{
    class Program
    {
        const int MAX_LICZBA_ELEMENTOW = 100;

        static void Main(string[] args)
        {
            var rand = new Random();

            Console.Write("Podaj liczbe elementow do posortowania (max 100): ");
            int liczbaElementow = int.Parse(Console.ReadLine());

            if (liczbaElementow > MAX_LICZBA_ELEMENTOW || liczbaElementow < 1)
            {
                Console.WriteLine("Nieprawidlowa liczba elementow. Zakres: 1-100.");
                return;
            }

            Console.Write("Czy chcesz wprowadzic liczby recznie? (tak/nie): ");
            string metodaGenerowania = Console.ReadLine();

            List<int> listaWartosci = new List<int>(liczbaElementow);

            if (metodaGenerowania.ToLower() == "tak")
            {
                for (int i = 0; i < liczbaElementow; i++)
                {
                    Console.Write("Podaj wartosc dla elementu " + (i + 1) + ": ");
                    listaWartosci.Add(int.Parse(Console.ReadLine()));
                }
            }
            else
            {
                for (int i = 0; i < liczbaElementow; i++)
                {
                    listaWartosci.Add(rand.Next(1, 101));
                }
            }

            Console.Write("Wybierz algorytm sortowania (babelkowe/wybieranie/wstawianie): ");
            string algorytm = Console.ReadLine();

            if (algorytm.ToLower() == "babelkowe")
            {
                SortowanieBabelkowe(listaWartosci);
            }
            else if (algorytm.ToLower() == "wybieranie")
            {
                SortowaniePrzezWybieranie(listaWartosci);
            }
            else if (algorytm.ToLower() == "wstawianie")
            {
                SortowaniePrzezWstawianie(listaWartosci);
            }
            else
            {
                Console.WriteLine("Nieznany algorytm sortowania.");
                return;
            }
        }

        static void SortowanieBabelkowe(List<int> lista)
        {
            int n = lista.Count;
            for (int i = 0; i < n; i++)
            {
                for (int j = 0; j < n - i - 1; j++)
                {
                    if (lista[j] > lista[j + 1])
                    {
                        int temp = lista[j];
                        lista[j] = lista[j + 1];
                        lista[j + 1] = temp;
                    }
                }
                Wyswietl(lista);
            }
        }

        static void SortowaniePrzezWybieranie(List<int> lista)
        {
            int n = lista.Count;
            for (int i = 0; i < n - 1; i++)
            {
                int minIndeks = i;
                for (int j = i + 1; j < n; j++)
                {
                    if (lista[j] < lista[minIndeks])
                    {
                        minIndeks = j;
                    }
                }
                int temp = lista[minIndeks];
                lista[minIndeks] = lista[i];
                lista[i] = temp;
                Wyswietl(lista);
            }
        }

        static void SortowaniePrzezWstawianie(List<int> lista)
        {
            int n = lista.Count;
            for (int i = 1; i < n; i++)
            {
                int klucz = lista[i];
                int j = i - 1;
                while (j >= 0 && lista[j] > klucz)
                {
                    lista[j + 1] = lista[j];
                    j = j - 1;
                }
                lista[j + 1] = klucz;
                Wyswietl(lista);
            }
        }

        static void Wyswietl(List<int> lista)
        {
            foreach (var item in lista)
            {
                Console.Write(item + " ");
            }
            Console.WriteLine();
        }
    }
}

File: lekcje/01_algorytmy_i_zlozonosc/przyklady/sortowanie.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define MAX_LICZBA_ELEMENTOW 100

void sortowanie_babelkowe(int *lista, int n);
void sortowanie_przez_wybieranie(int *lista, int n);
void sortowanie_przez_wstawianie(int *lista, int n);
void wyswietl(int *lista, int n);

int main() {
  srand(time(0));

  int liczba_elementow;
  char algorytm[20];
  char metoda_generowania[4];

  printf("Podaj liczbe elementow elementow do posortowania (max 100): ");
  scanf("%d", &liczba_elementow);

  if (liczba_elementow > MAX_LICZBA_ELEMENTOW || liczba_elementow < 1) {
    printf("Nieprawidlowa liczba elementow. Zakres: 1-100.\n");
    return 1;
  }

  printf("Czy chcesz wprowadzic liczby recznie? (tak/nie): ");
  scanf("%s", metoda_generowania);

  int lista_wartosci[liczba_elementow];

  if (strcmp(metoda_generowania, "tak") == 0) {
    for (int i = 0; i < liczba_elementow; i++) {
      printf("Podaj wartosc dla elementu %d: ", i + 1);
      scanf("%d", &lista_wartosci[i]);
    }
  } else {
    for (int i = 0; i < liczba_elementow; i++) {
      lista_wartosci[i] = rand() % 100 + 1;
    }
  }

  printf("Wybierz algorytm sortowania (babelkowe/wybieranie/wstawianie): ");
  scanf("%s", algorytm);

  if (strcmp(algorytm, "babelkowe") == 0) {
    sortowanie_babelkowe(lista_wartosci, liczba_elementow);
  } else if (strcmp(algorytm, "wybieranie") == 0) {
    sortowanie_przez_wybieranie(lista_wartosci, liczba_elementow);
  } else if (strcmp(algorytm, "wstawianie") == 0) {
    sortowanie_przez_wstawianie(lista_wartosci, liczba_elementow);
  } else {
    printf("Nieznany algorytm sortowania.\n");
    return 1;
  }

  return 0;
}

void sortowanie_babelkowe(int *lista, int n) {
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n - i - 1; j++) {
      if (lista[j] > lista[j + 1]) {
        int temp = lista[j];
        lista[j] = lista[j + 1];
        lista[j + 1] = temp;
      }
    }
    wyswietl(lista, n);
  }
}

void sortowanie_przez_wybieranie(int *lista, int n) {
  for (int i = 0; i < n - 1; i++) {
    int min_indeks = i;
    for (int j = i + 1; j < n; j++) {
      if (lista[j] < lista[min_indeks]) {
        min_indeks = j;
      }
    }
    int temp = lista[min_indeks];
    lista[min_indeks] = lista[i];
    lista[i] = temp;
    wyswietl(lista, n);
  }
}

void sortowanie_przez_wstawianie(int *lista, int n) {
  for (int i = 1; i < n; i++) {
    int klucz = lista[i];
    int j = i - 1;
    while (j >= 0 && lista[j] > klucz) {
      lista[j + 1] = lista[j];
      j = j - 1;
    }
    lista[j + 1] = klucz;
    wyswietl(lista, n);
  }
}

void wyswietl(int *lista, int n) {
  for (int i = 0; i < n; i++) {
    printf("%d ", lista[i]);
  }
  printf("\n");
}
